Assembler (#7)
jthulhu/cpu: docs/instructionset.tex

\documentclass{scrartcl}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage[english]{babel}
\usepackage{amsmath}
\usepackage{booktabs, array}
\usepackage{tikz, environ}
\usepackage{pgffor, fp}
\usepackage{hyperref}

%%%%%%%%%
% SETUP %
%%%%%%%%%

\usetikzlibrary{decorations.pathreplacing, calligraphy, arrows.meta}

\DeclareMathOperator{\op}{op}
\DeclareMathOperator{\instr}{instr}

\makeatletter
\newsavebox{\measure@tikzpicture}
\NewEnviron{scalepicture}[1]{%
  \def\tikz@width{#1}%
  \def\tikzscale{1}\begin{lrbox}{\measure@tikzpicture}%
  \BODY
  \end{lrbox}%
  \pgfmathparse{#1/\wd\measure@tikzpicture}%
  \edef\tikzscale{\pgfmathresult}%
  \BODY
}
\makeatother

%%%%%%%%%%%%
% DOCUMENT %
%%%%%%%%%%%%

\title{Instruction set reference}
\author{%
  Jean \textsc{Caspar},
  Loïc \textsc{Chevalier},
  Vladimir \textsc{Ivanov},
  Adrien \textsc{Mathieu}
}
\date{}

\begin{document}
\maketitle{}
\section*{Hardware requirements}
This instruction set requires 16 registers, \(r_0,\ldots,r_{15}\).\par
The first register always contains zeros. It can be used as any other register.\par

> All the other registers are general purpose 16 bit register, which can be

Reviewer: ```suggestion
All the other registers are general purpose 16 bit registers, which can be
```

accessed both for writing and for reading, even in a single cycle, in which case
the data read is the one stored in the register \textit{before} the write

> operation. Every register must initially be initialized to zeros, as should the

Reviewer: ```suggestion
operation. At startup, every register must be reset to contain only zeros, as should the
```

Author: J'ai légèrement changé ta formulation.

Reviewer: Il n'y a aucun changement avec l'ancienne version... (mon problème étant le "**initially** be **initialized**")

Author: Il y a visiblement eu un problème, je sais pas trop où est fini la révision.

Author: "Initially, every register must be set to zero" ça te va?

Reviewer: oui, c'est bon aussi

Reviewer: EDIT : "including the instruction pointer register" aussi je pense qu'il faut rajouter

Author: C'est fait.

RAM.

\section*{Instructions}

> An instruction \verb|instr| takes 32 bits. The first eight bits encode the

Reviewer: ```suggestion
An instruction is 32 bits wide. The first eight bits encode the
```

A $0$ means an arity of one, a $1$ means an arity of two. Every instruction
additionally receives a \emph{destination register} $r_d$, in which the result of
the operation must be stored, unless told otherwise.\par
The next five bits encode to the operation. The first two bits indicate the
operation family:
\begin{itemize}
\item $00$: Arithmetic and logic operation.
\item $01$: Jump operation.
\item $10$: RAM operation
\item $11$: Comparison operation.
\end{itemize}
The next three bits specify the actual operation performed.\par
The two final bits of the instruction code indicate whether arguments are
registers or immediates, $1$ meaning register and $0$ immediate. The first of
these two bits corresponds to $a_1$, whereas the second to $a_2$.
\subsection*{Arguments}
Arguments are of two kind: immediates and registers. Registers require four bits,
whereas immediates are 16 bits signed integers. At most one immediate can be
given.\par
The first ``argument'' is always a register, $r_d$. The next argument $a_1$, or
the next two arguments $a_1$ and $a_2$, in this order, can be registers or
immediates.

\section*{Implementation details}
\subsection*{RAM operations}

> \verb|load| and \verb|store| are encoded in the same way, but for the arity. This

Reviewer: ```suggestion
\verb|load| and \verb|store| are encoded in the same way, except for the arity. This
```

      \node at (2,-4.5) {$r_2$};
      \foreach \n in {4,...,15}{
        \node at (\n.5,-3.5) {X};
      }

      %% 1, 10
      \foreach \n in {0,...,31}{
        \FPeval{\next}{clip(\n+1)}
        \draw (\n,-6) rectangle (\next,-5);
      }
      \foreach \n in {32,31,26,24,20,16,0}{
        \draw[dotted] (\n,-7) -- (\n,-6);
      }
      \node at (28.5,-6.5) {operation};
      \node at (22,-6.5) {$r_d$};
      \node at (31.5,-5.5) {$1$};
      
      \node at (25.5,-5.5) {$1$};
      \node at (24.5,-5.5) {$0$};
      \node at (18,-6.5) {$r_1$};
      \node at (8,-6.5) {$i$};

      %% 1, 01
      \foreach \n in {0,...,31}{
        \FPeval{\next}{clip(\n+1)}
        \draw (\n,-8) rectangle (\next,-7);
      }
      \foreach \n in {32,31,26,24,20,4,0}{
        \draw[dotted] (\n,-9) -- (\n,-8);
      }
      \node at (28.5,-8.5) {operation};
      \node at (22,-8.5) {$r_d$};
      \node at (31.5,-7.5) {$1$};
      
      \node at (25.5,-7.5) {$0$};
      \node at (24.5,-7.5) {$1$};
      \node at (12,-8.5) {$i$};
      \node at (2,-8.5) {$r_2$};
    \end{tikzpicture}
  \end{scalepicture}
  \caption{Layout of instructions}
\end{figure}

\begin{table}
  \center
  \label{tab:instrs}

>   \caption{Layout of instructions}

Reviewer: This must be rewritten (if I understood correctly, `instr[0:0]` is actually `instr[31:31]`, etc)

Author: Ce sera changé au prochain commit.

Author: Normalement actuellement ça va.

Reviewer: pareil, pas de changement, tu as oublié de `git add` peut-être ?

Author: Heu oui je sais pas trop ce que j'ai fait, je corrige ça.

Author: Ah oui non attends, ça ça allait, ça n'a pas changé. Si tu regardes le dessin, l'arité c'est encore `instr[0:0]`.

Reviewer: Si tu le dis, mais dans ce cas il faut corriger le deuxième bloc, il y a un `instr[5:2]` à la place d'un `instr[5:1]`

Author: Ah oui exact, good catch.

Author: Ok c'est bon.
